src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs

```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Roslyn.Utilities
+{
+    /// <summary>
+    /// A set that returns the inserted values in insertion order.
+    /// The mutation operations are not thread-safe.
+    /// </summary>
+    internal sealed class SetWithInsertionOrder<T> : IEnumerable<T>
+    {
+        private Dictionary<T, uint> _map = new Dictionary<T, uint>();
+        private uint _nextElementValue = 0;
+        private T[] _elements = null;
+
+        public bool Add(T value)
+        {
+            if (_map.ContainsKey(value)) return false;
+            var thisValue = _nextElementValue++;
+            _map.Add(value, thisValue);
+            if (_elements == null)
+            {
+                _elements = new T[10];
```

Comment:
I stuck my fingers in the air and felt which way the wind was blowing. I felt the wind on all ten fingers.